### ISSUE TITLE
acknowledgeSystemAlert breaks view hierarchy 

### ISSUE THREAD

**Issue by foulkesjohn** [kif-framework/KIF#678]
I have a test which uses `tapViewWithAccessibilityLabel:value:traits:` to tap a `UIButton` which is inside a `UICollectionViewCell`. This was working fine until I added a call to `acknowledgeSystemAlert` to dismiss the alert which is displayed for push notifications at the start of my app running. If I disable this call then it works again.

I have dug around and found that when the call to `acknowledgeSystemAlert` is made when trying to find the button to tap the buttons superview becomes an instance `_UISnapshotWindow` and therefore when `viewContainingAccessibilityElement:tappable:error:` iterates through the view hierarchy it is unable to find the `UICollectionView` and scroll to the cell containing the button.

Any ideas how to work around this?

**Comment by jessealva** [comment]
So i actually found what i believe is a bug in the implementation of the acknowledgeSystemAlert function.  I have two alerts which I need to dismiss, one for notifications and one for Location awareness.  it seemed that the function would just hang forever after dismissing the first alert. I changed the following method from this:

- (void)acknowledgeSystemAlert {
    UIAApplication *application = [[self target] frontMostApp];
    UIAAlert *alert = application.alert;

    if (![alert isKindOfClass:[self nilElementClass]]) {
        [[alert.buttons lastObject] tap];
        while (![application.alert isKindOfClass:[self nilElementClass]]) {}
    }
}

to this..

- (void)acknowledgeSystemAlert {
    UIAApplication *application = [[self target] frontMostApp];
    UIAAlert *alert = application.alert;

    if (![alert isKindOfClass:[self nilElementClass]]) {
        [[alert.buttons lastObject] tap];
        while (![application.alert isKindOfClass:[self nilElementClass]]) {
            alert = application.alert;
            [[alert.buttons lastObject] tap];
        }
    }
}

And that seemed to clear up my infinite hang.  Not sure if this is your problem or not.  Once i added this code, i was able to call acknowledgeSystemAlert() to dismiss ALL my alerts, and then I could continue running my tests as desired.

**Comment by phatmann** [comment]
@jessealva, really glad to see a fix for this. Care to submit a pull request?

**Comment by phatmann** [comment]
There is now an open PR to fix this: #691.

**Comment by foulkesjohn** [comment]
This doesn't fix the original issue. This issue was hijacked by @jessealva (no offence). Using `acknowledgeSystemAlert` causes the view hierarchy to go haywire

**Comment by phatmann** [comment]
@foulkesjohn I will reopen this issue. Do you have any ideas for a fix?

**Comment by foulkesjohn** [comment]
No right now. I have a few other avenues to explore to try and figure out what the issue is but no immediate plans to look into it. I managed to work around this for a while but Xcode 7/iOS 9 system alert views seem to block alerts displayed by my app. Which is why I revisited this issue.

**Comment by foulkesjohn** [comment]
I've managed to reproduce a similar issue when using `acknowledgeSystemAlert` in this repo: https://github.com/foulkesjohn/kiftest.

There is a difference in behaviour when `acknowledgeSystemAlert` is called before tests. Without calling it the buttons selection state is retained, but when it is called it is reset by `awakeFromNib` in `HeaderView`. This isn't the original issue I described but I believe its similar. I have yet to find a workaround for whatever `acknowledgeSystemAlert` does to the view hierarchy

**Comment by phatmann** [comment]
What do you think about saving off the view hierarchy and then restoring it as a workaround?

**Comment by foulkesjohn** [comment]
I'm not sure what you mean, how would that work?

**Comment by phatmann** [comment]
I had not thought about the mechanics yet. If this call really trashes the view hierarchy, don't we need to somehow restore it? Otherwise what other fix could we use?

**Comment by foulkesjohn** [comment]
I don't know enough about what linking the UIAutomation framework does to the app to try and fix it or offer any kind of advice unfortunately. I intend to change my tests to deal with this as a workaround for the time being.
